packages/babel-plugin-proposal-import-wasm-source/src/index.ts

```diff
@@ -0,0 +1,309 @@
+    }
+    throw new Error(
+      `@babel/plugin-proposal-import-wasm-source can only be used when not ` +
+        `compiling modules, or when compiling them to CommonJS.`,
+    );
+  };
+
+  return {
+    name: "proposal-import-wasm-source",
+
+    inherits: syntaxImportSourcePhase,
+
+    visitor: {
+      Program(path) {
+        if (path.node.sourceType !== "module") return;
+
+        const helper = getHelper(this.file);
+
+        // TS reports that 'Assertions require every name in the call target to be
+        // declared with an explicit type annotation.' if we just call
+        // t.assertImportDefaultSpecifier, so we _explicitly_ annotate `t` as
+        // having type `typeof t`. Unfortunately this is a design limitation of
+        // the TS type checker, so we need to manually help it:
+        // https://github.com/microsoft/TypeScript/issues/36931
+        const t2: typeof t = t;
+
+        const ids: t.Identifier[] = [];
+        const fetches: t.Expression[] = [];
+        for (const decl of path.get("body")) {
+          if (!decl.isImportDeclaration({ phase: "source" })) continue;
+
+          if (decl.node.attributes?.length || decl.node.assertions?.length) {
+            throw path.buildCodeFrameError(
+              "`import source` with import attributes cannot be compiled.",
+            );
+          }
+
+          const specifier = decl.node.specifiers[0];
+          t2.assertImportDefaultSpecifier(specifier);
+
+          ids.push(specifier.local);
+          fetches.push(helper.buildFetch(decl.node.source, path));
+          decl.remove();
+        }
+        if (ids.length === 0) return;
+
+        const declarators: t.VariableDeclarator[] = [];
+        if (ids.length === 1) {
+          let rhs = fetches[0];
+          if (helper.needsAwait) rhs = t.awaitExpression(rhs);
+          declarators.push(t.variableDeclarator(ids[0], rhs));
+        } else if (helper.needsAwait) {
+          declarators.push(
+            t.variableDeclarator(
+              t.arrayPattern(ids),
+              t.awaitExpression(
+                template.expression.ast`
+                  Promise.all(${t.arrayExpression(fetches)})
+                `,
+              ),
+            ),
+          );
+        } else {
+          for (let i = 0; i < ids.length; i++) {
+            declarators.push(t.variableDeclarator(ids[i], fetches[i]));
+          }
+        }
+
+        path.unshiftContainer(
+          "body",
+          t.variableDeclaration("const", declarators),
+        );
+      },
+
+      ImportExpression(path) {
+        if (path.node.phase !== "source") return;
+
+        if (path.node.options) {
+          throw path.buildCodeFrameError(
+            "`import.source` with an options bag cannot be compiled.",
+          );
+        }
+
+        path.replaceWith(
+          getHelper(this.file).buildFetchAsync(path.node.source, path),
+        );
+      },
+    },
+  };
+
+  function buildHelper(toCommonJS: boolean) {
```

Comment:
https://github.com/babel/babel/pull/15709/files#diff-4e18d6bcb1580209822d484893ec4ad14961e07aaf2f918e7d43bb6081e60ecfR351-R372

Can we do something similar?

Note: We always need to emit `import.meta.resolve("./x.wasm")` for the bundler.

Reply:
Given that it's very likely that files would include just one `.wasm` import (I have never seen even whole packages with two `.wasm` entry points :) ), in the general case this would just make the code slightly bigger.

Reply:
Btw, I would expect the community to come up with their own bundler-specific plugins, such as the webpack output I showed above. It makes it possible to have much more bundler-integrated support!